### ISSUE TITLE
TensorFlowException: Op type not registered 'NonMaxSuppressionV3' in binary running on localhost.

### ISSUE THREAD

**Issue by japer21** [tensorflow/tensorflow#19854]
```
### System information
Linux 4.4.0-127-generic #153-Ubuntu SMP Sat May 19 10:58:46 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
VERSION="16.04.4 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

== compiler =====================================================
c++ (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux 4.4.0-127-generic #153-Ubuntu SMP Sat May 19 10:58:46 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy                          1.14.2                
protobuf                       3.5.2.post1           
tensorflow                     1.8.0                 
tensorflow-serving-api-python3 1.4.0                 

== check for virtualenv =========================================
True

== tensorflow import ============================================
tf.VERSION = 1.8.0
tf.GIT_VERSION = v1.8.0-0-g93bc2e2072
tf.COMPILER_VERSION = v1.8.0-0-g93bc2e2072
Sanity check: array([1], dtype=int32)

- **Exact command to reproduce**: I run my android application on Android Studio
```

### Describe the problem
I have an android application where for instance I make an initialization of my frozen model. 
It seems that there is a problem with the ops. I get this exception:

`java.lang.RuntimeException: Unable to start activity ComponentInfo{myapp/myapp.MainActivity}: 
org.tensorflow.TensorFlowException: Op type not registered 'NonMaxSuppressionV3' in binary running on localhost. Make sure the Op and Kernel are registered in the binary running in this process.

        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2666)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2727)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1478)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6121)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:890)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:780)
     Caused by: org.tensorflow.TensorFlowException: Op type not registered 'NonMaxSuppressionV3' in binary running on localhost. Make sure the Op and Kernel are registered in the binary running in this process.
        at org.tensorflow.Graph.importGraphDef(Native Method)
        at org.tensorflow.Graph.importGraphDef(Graph.java:130)
        at org.tensorflow.Graph.importGraphDef(Graph.java:114)
        at org.tensorflow.contrib.android.TensorFlowInferenceInterface.loadGraph(TensorFlowInferenceInterface.java:559)
        at org.tensorflow.contrib.android.TensorFlowInferenceInterface.<init>(TensorFlowInferenceInterface.java:105)
        at myapp.ActivityInference.<init>(ActivityInference.java:22)
        at myapp.MainActivity.onCreate(MainActivity.java:21)
        at android.app.Activity.performCreate(Activity.java:6692)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2619)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2727) 
        at android.app.ActivityThread.-wrap12(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1478) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.app.ActivityThread.main(ActivityThread.java:6121) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:890) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:780) ` 

I have checked the BUILD file in tensorflow/core/kernels/ and there is defined the kernel library:
`tf_kernel_library(
    name = "non_max_suppression_op",
    prefix = "non_max_suppression_op",
    deps = IMAGE_DEPS,
)`
I am new to this so I really do not know how I can fix it.
I am working on a pre-trained frozen model.

### Source code / logs
Android code:

```
package myapp;
import android.content.Context;
import android.content.res.AssetManager;

import org.tensorflow.contrib.android.TensorFlowInferenceInterface;


public class ActivityInference {

    private static ActivityInference activityInferenceInstance;
    private TensorFlowInferenceInterface inferenceInterface;
    private static AssetManager assetManager;
    private static final String MODEL_FILE = "file:///android_asset/frozen_inference_graph.pb";
    private static final String INPUT_NODE = "input";

    public ActivityInference(final Context context) {
        this.assetManager = context.getAssets();
        inferenceInterface = new TensorFlowInferenceInterface(assetManager, MODEL_FILE);

    }

    public static ActivityInference getInstance(final Context context){
        if (activityInferenceInstance == null)
        {
            activityInferenceInstance = new ActivityInference(context);
        }
        return activityInferenceInstance;
    }
}
```

and my main activity:

```
package myapp;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;


public class MainActivity extends AppCompatActivity {

    private ActivityInference activityInference;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        final TextView textView = (TextView) findViewById(R.id.textView);
        activityInference = new ActivityInference(getApplicationContext());
    }
}
```

**Comment by japer21** [comment]
@andrewharp  Would you mind to see this issue I raised please? I really need some help. Thanks.

**Comment by achraf-boussaada** [comment]
you should basically add the missing op and rebuild the library manually as mentioned in the issue you referenced. Did you manage to do it ? I'm facing the same problem with the same missing op after training my network with the `mobilenet V2 SSDlite`

**Comment by japer21** [comment]
@achraf-boussaada  No I did not, when I do it with tflite it works but with mobile version of mobile it does not, and I can not rebuild the library.

**Comment by achraf-boussaada** [comment]
Well I managed to build the library and use it with `Tensorflow mobile` but the results were disappointing (almost 2 seconds inference for object detection). Did you get better results when using `tflite` ? 

to get it to work on `Tensorflow mobile` I cloned `Tensorflow` with this line to avoid protobuf compilation errors: 
`git clone --recurse-submodules https://github.com/tensorflow/tensorflow.git `

then proceeded normally building the library from source and followed the instruction used in the issue you referenced and it worked for me.

**Note :** create a `virtualenv` and clone and build `tensorflow` from there, it isolates it and help avoid any conflict problems.

**Comment by japer21** [comment]
@achraf-boussaada  Well I am working on windows, so it seems a little complicated. 
And I am trying to include my model mobile to be able to do the TF Detect but I get this error, although I have defined cmake in the gradle build file.
Yes the mobile version seems slower than the tflite of course. But the confidence is less when it'a about the tflite.

**Comment by japer21** [comment]
@achraf-boussaada  You have your own model mobile or you are using their model? When it's with their model it works perfectly, but when I put my mobile model it does not do the job and raises this exception.

**Comment by achraf-boussaada** [comment]
@japer21  hmm ! good to know. I took the `mobilenet V2 SSDlite` checkpoint and I trained my custom model with it. I had also disappointing results when I used `tflite` with a model trained with `Mobilenet V1 SSD` (inference time was way better with Tensorflow mobile- not only me encountered this issue-). That's why I'm holding off about using `tflite` and I'm waiting until it's stable enough to switch.

**Comment by japer21** [comment]
@achraf-boussaada You use android on which OS system please? I am on Windows and with the most recent and stable version of andrfoid tensorflow it does not work.
Would you mind explaining me a little more of the steps you took to make it working after training your model? Thanks in advance.

**Comment by japer21** [comment]
@achraf-boussaada In fact, the model I am using has been trained with Inception and not mobilenet V2, maybe that can cause also a problem.

**Comment by achraf-boussaada** [comment]
frist of all it doesn't matter which OS you're using when working with android. It only matters for model training, afterwards you're going to freeze your model and add that file to your android project. 

I'm using the [Object detection demo](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android) from tensorflow because I need to detect objects in realtime, so all I had to do is add my new model to the asset folder and do minor changes in the code. As it seems from your code snippet you're trying to write an app from scratch and since you're using Inception I guess you're trying to classify images. So take a look at the demo I referenced you can test your model there and have a good idea how things by classification works.

**Comment by japer21** [comment]
@achraf-boussaada Did you change something to make it work without giving the exception `org.tensorflow.TensorFlowException: Op type not registered 'NonMaxSuppressionV3' in binary running on localhost. Make sure the Op and Kernel are registered in the binary running in this process.`  ?

**Comment by achraf-boussaada** [comment]
I built the library manually for different cpu architectures as I mentioned in my earlier comment to add the missing `op`, after that i added the library to my project. my project structure looks like this : 
(the libraries should be placed in the `libs` directory in your project)

> --jniLibs
> ----arm64-v8
> --------libtensorflow_inference.so
> ----armeabi-v7a
> --------libtensorflow_inference.so
> ----x86
> --------libtensorflow_inference.so
> ----x86_64
> --------libtensorflow_inference.so
> ----libandroid_tensorflow_inference_java.jar

you should download the `libandroid_tensorflow_inference_java.jar` from the last successful [nightlybuild](https://ci.tensorflow.org/view/Nightly/job/nightly-android/) referenced in their Object Detection Demo Readme but I don't know why for about a month now the link don't work anymore. Try to open a new issue about it, because I can't attach it here  and I don't know if there is a new link for it.

you need to add this to your gradle file : 

```
android{
...
    sourceSets {
        main {
            jniLibs.srcDirs = ['libs']
        }
    }
...
}
```
all of this will basically replace this line : 
`implementation 'org.tensorflow:tensorflow-android:+'`
so after adding the libraries delete that line to avoid conflict. And that should be it

**Comment by japer21** [comment]
@achraf-boussaada  This link https://ci.tensorflow.org/view/Nightly/job/nightly-android/ seems broken as it gives me 404 Error. I will try these steps thanks

**Comment by achraf-boussaada** [comment]
yep it's broken. It worked a month ago and in all documentations I saw, they still referencing that link. Maybe they restructured their Jenkins. Good luck and keep me updated , maybe I can help.

**Comment by japer21** [comment]
@achraf-boussaada You did it using these commands?

```
An example of command-line usage is:
  bazel build tensorflow/python/tools:print_selective_registration_header && \
  bazel-bin/tensorflow/python/tools/print_selective_registration_header \
    --graphs=path/to/graph.pb > ops_to_register.h

Then when compiling tensorflow, include ops_to_register.h in the include search
path and pass -DSELECTIVE_REGISTRATION and -DSUPPORT_SELECTIVE_REGISTRATION
 - see core/framework/selective_registration.h for more details.

When compiling for Android:
  bazel build -c opt --copt="-DSELECTIVE_REGISTRATION" \
    --copt="-DSUPPORT_SELECTIVE_REGISTRATION" \
    //tensorflow/contrib/android:libtensorflow_inference.so \
    --host_crosstool_top=@bazel_tools//tools/cpp:toolchain \
    --crosstool_top=//external:android/crosstool --cpu=armeabi-v7a
```

**Comment by achraf-boussaada** [comment]
yep ! I used those commands. I got an error at first but when I cloned the project using this : 
`git clone --recurse-submodules https://github.com/tensorflow/tensorflow.git`
it resolved my issue and it worked. Make sure to compile the library for all cpu architectures.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 119 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by petewarden** [comment]
I'm closing this issue since it's not been updated for a while, and seems to be fixed from the latest comments.
